File: node/mainImages.js
const { Worker } = require('worker_threads');
const config = require('./config');

/**
 * Parallel Batch Processing in Node (mainImages)
 * @author Marco Guillén <mguillen.developer@gmail.com>
 */
async function mainImages() {
    function worker(images) {
        return new Promise((resolve, reject) => {
            const worker = new Worker(config.workerFileImages, { workerData: images });
            worker.on(config.workerEventMessage, resolve);
            worker.on(config.workerEventError, reject);
        });
    }

    const start = Date.now();
    const segmentsPerWorker = Math.ceil(config.files.length / config.workersForImages);
    const groupWorkers = Array(config.workersForImages).fill().map((_, index) => {
        let images;
        if (index === 0) {
            images = config.files.slice(0, segmentsPerWorker);
        } else if (index === config.workersForImages - 1) {
            images = config.files.slice(segmentsPerWorker * index);
        } else {
            images = config.files.slice(segmentsPerWorker * index, segmentsPerWorker * (index + 1))
        }
        return worker(images)
    });
    Promise.all(groupWorkers).then(_ => console.log(`::: INTENSIVE JOB SUCCESSFULLY COMPLETED -> Workers: ${config.workersForImages} -> Elapsed time: ${Date.now() - start} ms`)); // Wait for Workers and Promises Group to finish
}

mainImages();

File: scala/src/main/scala/com/latamautos/challenge/benchmark/MainImages.scala
package com.latamautos.challenge.benchmark

import com.latamautos.challenge.benchmark.Config._
import java.io.{File, FileInputStream}
import com.sksamuel.scrimage.ImmutableImage
import com.sksamuel.scrimage.nio.JpegWriter
import scala.concurrent._
import java.util.concurrent.Executors

/**
 * Parallel Batch Processing in Scala (MainImages)
 * @author Marco Guillén <mguillen.developer@gmail.com>
 */
object MainImages {
  implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(images.size))

  def main(args: Array[String]) = {
    def worker(images: List[File]) = {
      images.foreach(i => ImmutableImage.loader().fromStream(new FileInputStream(i)).fit(X, Y).output(new JpegWriter().withCompression(Quality), Config.imageWriter(i)))
      printf("+++ WORKER FINISHED (processed images: %d)\n", images.size)
    }

    val start = System.currentTimeMillis
    val groupWorkers =  Future.sequence(images.grouped((images.size / WorkersForImages.toFloat).ceil.toInt).map(x => Future(worker(x))))
    groupWorkers.onComplete(result => printf(s"::: INTENSIVE JOB SUCCESSFULLY COMPLETED -> Workers: ${result.getOrElse(Nil).size} -> Elapsed time: ${System.currentTimeMillis - start} ms")) // Wait for Futures Group to finish
  }
}